text/0000-bee-model-crate/0000-bee-model-crate.md

```diff
@@ -0,0 +1,277 @@
+now so that we can flesh them out during the implementation phase or in future
+RFCs.
+
+```rust
+pub struct Trit(u8);
+pub struct SignatureOrMessageFragment([Trit; 6561]);
+pub struct Address([Trit; 243]);
+pub struct Tag([Trit; 81]);
+pub struct Timestamp(u64);
+pub struct Index(u64);
+pub struct Value(i64);
+pub struct BundleHash([Trit; 243]);
+pub struct TransactionHash([Trit; 243]);
+pub struct Nonce([Trit; 81]);
+```
+
+The `Transaction` type only contains two constructor methods getter methods for
+read-only borrows of its fields.
+
+```rust
+impl Transaction {
+    /// Create a `Transaction` from a reader.
+    pub fn from_reader<R: std::io::Read>(reader: R) -> Result<Self, TransactionError> {
+        unimplemented!()
+    }
+
+    /// Create a `Transaction` from a slice of bytes
+    pub fn from_slice(stream: &[u8]) -> Result<Self, TransactionError> {
+        unimplemented!()
+    }
+
+    pub fn signature_or_message_fragment(&self) -> &SignatureOrMessageFragment {
+        &self.signature_fragments
+    }
+
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    pub fn value(&self) -> &Value {
+        &self.value
+    }
+
+    ...
+}
+```
+
+The `from_slice` and `from_reader` methods will use a `TransactionDraft` internally,
+but these are implementation details.
+
+The `TransactionError` is not fully fleshed out yet. For the time being, it
+definitely contains `io::Error`:
+
+```rust
+pub enum TransactionError {
+    Io(io::Error),
+}
+```
+
+### `TransactionDraft`
+
+`TransactionDraft` is a basic builder pattern for creating and setting the individual
+fields of a new transaction.
+
+```rust
+#[derive(Default)]
+struct TransactionDraft {
+    signature_or_message_fragment: Option<SignatureOrMessageFragment>,
+    address: Option<Address>,
+    value: Option<Value>,
+    obsolete_tag: Option<ObsoleteTag>,
+    current_index: Option<CurrentIndex>,
+    last_index: Option<LastIndex>,
+    bundle_hash: Option<BundleHash>,
+    trunk: Option<Trunk>,
+    branch: Option<Branch>,
+    tag: Option<Tag>,
+    attachment_timestamp: Option<AttachmentTimestamp>,
+    attachment_timestamp_lower_bound: Option<AttachmentTimestampLowerbound>,
+    attachment_timestamp_upper_bound: Option<AttachmentTimestampUpperbound>,
+    nonce: Option<Nonce>
+}
+
+impl TransactionDraft {
+    /// Create a `TransactionDraft` from a reader.
+    pub fn from_reader<R: std::io::Read>(reader: R) -> Result<Self, TransactionDraftError> {
+        unimplemented!()
+    }
+
+    /// Create a `TransactionDraft` from a slice of bytes
+    pub fn from_slice(stream: &[u8]) -> Result<Self, TransactionDraftError> {
+        unimplemented!()
+    }
+
+    /// Verify that all fieleds are set and build the `Transaction`.
+    pub fn build(self) -> Result<Transaction, TransactionDraftError> {
+        unimplemented!()
+    }
+
+    pub fn address(&mut self, address: Address) -> &mut Self {
```

Comment:
How about give a proper setter method prefix like `set_address`?

Reply:
Yes, as RFC #344 notes it:
[https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#gettersetter-apis](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#gettersetter-apis)